bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/validation/SyncTableValidationOnlyJob.java

```diff
@@ -0,0 +1,1102 @@
+    }
+
+    /**
+     * create connection to data store. if zkClusterConfKey is null or empty string, the connection
+     * will default to properties set in conf for establishing connection.
+     *
+     * @param conf configuration for connection
+     * @param zkClusterConfKey zookeeper configuration key for source or target
+     * @return Connection to cluster
+     * @throws IOException on connection
+     */
+    private static Connection openConnection(Configuration conf, String zkClusterConfKey)
+        throws IOException {
+      String zkCluster = conf.get(zkClusterConfKey);
+      Configuration clusterConf = HBaseConfiguration.createClusterConf(conf, zkCluster);
+      return ConnectionFactory.createConnection(clusterConf);
+    }
+
+    private static Table openTable(
+        Connection connection, Configuration conf, String tableNameConfKey) throws IOException {
+      return connection.getTable(TableName.valueOf(conf.get(tableNameConfKey)));
+    }
+
+    /**
+     * Attempt to read the next source key/hash pair. If there are no more, set nextSourceKey to
+     * null
+     */
+    private void findNextKeyHashPair() throws IOException {
+      boolean hasNext = sourceHashReader.next();
+      if (hasNext) {
+        nextSourceKey = sourceHashReader.getCurrentKey();
+      } else {
+        // no more keys - last hash goes to the end
+        nextSourceKey = null;
+      }
+    }
+
+    @Override
+    protected void map(ImmutableBytesWritable key, Result value, Context context)
+        throws IOException, InterruptedException {
+      try {
+        // first, finish any hash batches that end before the scanned row
+        while (nextSourceKey != null && key.compareTo(nextSourceKey) >= 0) {
+          moveToNextBatch(context);
+        }
+
+        // next, add the scanned row (as long as we've reached the first batch)
+        if (targetHasher.isBatchStarted()) {
+          targetHasher.hashResult(value);
+        }
+      } catch (Throwable t) {
+        mapperException = t;
+        Throwables.propagateIfInstanceOf(t, IOException.class);
+        Throwables.propagateIfInstanceOf(t, InterruptedException.class);
+        Throwables.propagate(t);
+      }
+    }
+
+    /**
+     * If there is an open hash batch, complete it and sync if there are diffs. Start a new batch,
+     * and seek to read the
+     */
+    private void moveToNextBatch(Context context) throws IOException, InterruptedException {
+      if (targetHasher.isBatchStarted()) {
+        finishBatchAndCompareHashes(context);
+      }
+      targetHasher.startBatch(nextSourceKey);
+      currentSourceHash = sourceHashReader.getCurrentHash();
+
+      findNextKeyHashPair();
+    }
+
+    /**
+     * Finish the currently open hash batch. Compare the target hash to the given source hash. If
+     * they do not match, then sync the covered key range.
+     */
+    private void finishBatchAndCompareHashes(Context context)
```

Comment:
I didn't read after this, I am not sure why we are re-writing this logic here? 

I want to understand why we can't use sync table code. Is there no way to "inject" Bigtable connection when target connection is created and just use that code. It seems like we are duplicating a whole lot of code from there, I am not sure what we are getting by copying it.